### ISSUE TITLE
Dispatcher stream error, tuner may have crashed

### ISSUE THREAD

**Issue by imsaumil** [microsoft/nni#5162]
**Describe the issue**:

**Environment**:
- NNI version: 2.9
- Training service (local|remote|pai|aml|etc): Local
- Client OS: Windows 10
- Server OS (for remote mode only):
- Python version: 3.7.13 (also tried with 3.9.12)
- PyTorch/TensorFlow version: torch 1.12.1+cu116
- Is conda/virtualenv/venv used?: Yes
- Is running in Docker?: No

**Configuration**:
 - Experiment config (remember to remove secrets!):
{
  "params": {
    "experimentType": "hpo",
    "trialCommand": "python model.py",
    "trialCodeDirectory": "C:\\Users\\**removed**",
    "trialConcurrency": 2,
    "maxTrialNumber": 50,
    "useAnnotation": false,
    "debug": false,
    "logLevel": "info",
    "experimentWorkingDirectory": "C:\\Users\\sashah8\\nni-experiments",
    "tuner": {
      "name": "TPE",
      "classArgs": {
        "optimize_mode": "maximize"
      }
    },
    "trainingService": {
      "platform": "local",
      "trialCommand": "python model.py",
      "trialCodeDirectory": "C:\\Users\\**removed*",
      "debug": false,
      "maxTrialNumberPerGpu": 1,
      "reuseMode": false
    }
  },
  "execDuration": "9s",
  "nextSequenceId": 2,
  "revision": 4
}

 - Search space:
 {
    'features': {'_type': 'choice', '_value': [128, 256, 512, 1024]},
    'lr': {'_type': 'loguniform', '_value': [0.0001, 0.1]},
    'momentum': {'_type': 'uniform', '_value': [0, 1]},
}

**Log message**:
- command prompt output:

[2022-10-16 11:19:45] Creating experiment, Experiment ID: dhe2x64y
[2022-10-16 11:19:45] Starting web server...
[2022-10-16 11:19:45] Setting up...
[2022-10-16 11:19:45] Web portal URLs: http://10.76.56.148:8057 http://169.254.31.88:8057 http://169.254.194.255:8057 http://169.254.35.0:8057 http://127.0.0.1:8057
Error: Dispatcher stream error, tuner may have crashed.                                                                                                            
    at EventEmitter.<anonymous> (C:\Users\sashah8\Anaconda3\envs\nni_HPO\lib\site-packages\nni_node\core\nnimanager.js:647:32)                                     
    at EventEmitter.emit (node:events:526:28)                                                                                                                      
    at WebSocketChannelImpl.handleError (C:\Users\sashah8\Anaconda3\envs\nni_HPO\lib\site-packages\nni_node\core\tuner_command_channel\websocket_channel.js:107:22)
    at WebSocketChannelImpl.heartbeat (C:\Users\sashah8\Anaconda3\envs\nni_HPO\lib\site-packages\nni_node\core\tuner_command_channel\websocket_channel.js:91:18)   
    at listOnTimeout (node:internal/timers:559:17)                                                                                                                 
    at processTimers (node:internal/timers:502:7)                                                                                                                  
[2022-10-16 11:20:32] Stopping experiment, please wait...
[2022-10-16 11:20:32] Experiment stopped
(nni_HPO) PS C:\Users\sashah8\CS - 791\HW3 - Hyperparameter Optimization> python main.py
[2022-10-16 11:20:56] Creating experiment, Experiment ID: xo1l8hi6
[2022-10-16 11:20:56] Starting web server...
[2022-10-16 11:20:57] Setting up...
[2022-10-16 11:20:59] Web portal URLs: http://10.76.56.148:8057 http://169.254.31.88:8057 http://169.254.194.255:8057 http://169.254.35.0:8057 http://127.0.0.1:8057
Error: Dispatcher stream error, tuner may have crashed.                                                                                                            
    at EventEmitter.<anonymous> (C:\Users\sashah8\Anaconda3\envs\nni_HPO\lib\site-packages\nni_node\core\nnimanager.js:647:32)
    at EventEmitter.emit (node:events:526:28)
    at WebSocketChannelImpl.handleError (C:\Users\sashah8\Anaconda3\envs\nni_HPO\lib\site-packages\nni_node\core\tuner_command_channel\websocket_channel.js:107:22)
    at WebSocketChannelImpl.heartbeat (C:\Users\sashah8\Anaconda3\envs\nni_HPO\lib\site-packages\nni_node\core\tuner_command_channel\websocket_channel.js:91:18)   
    at listOnTimeout (node:internal/timers:559:17)
    at processTimers (node:internal/timers:502:7)
AssertionError [ERR_ASSERTION]: Actual status: ERROR
    at NNIManager.manageTrials (C:\Users\sashah8\Anaconda3\envs\nni_HPO\lib\site-packages\nni_node\core\nnimanager.js:561:29)  
    at async Promise.all (index 3)
    at async NNIManager.run (C:\Users\sashah8\Anaconda3\envs\nni_HPO\lib\site-packages\nni_node\core\nnimanager.js:621:9) {    
  cause: AssertionError [ERR_ASSERTION]: Actual status: ERROR
      at NNIManager.manageTrials (C:\Users\sashah8\Anaconda3\envs\nni_HPO\lib\site-packages\nni_node\core\nnimanager.js:561:29)
      at async Promise.all (index 3)
      at async NNIManager.run (C:\Users\sashah8\Anaconda3\envs\nni_HPO\lib\site-packages\nni_node\core\nnimanager.js:621:9) {  
    generatedMessage: false,
    code: 'ERR_ASSERTION',
    actual: false,
    expected: true,
    operator: '=='
  }
}
[2022-10-16 11:25:29] Stopping experiment, please wait...
[2022-10-16 11:25:29] Experiment stopped

 - nnimanager.log: 
[2022-10-16 11:20:57] INFO (main) Start NNI manager
[2022-10-16 11:20:57] INFO (NNIDataStore) Datastore initialization done
[2022-10-16 11:20:57] INFO (RestServer) Starting REST server at port 8057, URL prefix: "/"
[2022-10-16 11:20:57] WARNING (NNITensorboardManager) Tensorboard may not installed, if you want to use tensorboard, please check if tensorboard installed.
[2022-10-16 11:20:57] INFO (RestServer) REST server started.
[2022-10-16 11:20:57] INFO (NNIManager) Starting experiment: xo1l8hi6
[2022-10-16 11:20:57] INFO (NNIManager) Setup training service...
[2022-10-16 11:20:57] INFO (LocalTrainingService) Construct local machine training service.
[2022-10-16 11:20:57] INFO (NNIManager) Setup tuner...
[2022-10-16 11:20:59] INFO (NNIManager) Change NNIManager status from: INITIALIZED to: RUNNING
[2022-10-16 11:21:00] INFO (NNIManager) Add event listeners
[2022-10-16 11:21:00] INFO (LocalTrainingService) Run local machine training service.
[2022-10-16 11:21:00] INFO (NNIManager) NNIManager received command from dispatcher: ID, 
[2022-10-16 11:21:00] INFO (NNIManager) NNIManager received command from dispatcher: TR, {"parameter_id": 0, "parameter_source": "algorithm", "parameters": {"features": 512, "lr": 0.00011599497432385631, "momentum": 0.5394097683782016}, "parameter_index": 0}
[2022-10-16 11:21:00] INFO (NNIManager) NNIManager received command from dispatcher: TR, {"parameter_id": 1, "parameter_source": "algorithm", "parameters": {"features": 128, "lr": 0.010676342217290587, "momentum": 0.5847482630995584}, "parameter_index": 0}
[2022-10-16 11:21:05] INFO (NNIManager) submitTrialJob: form: {
  sequenceId: 0,
  hyperParameters: {
    value: '{"parameter_id": 0, "parameter_source": "algorithm", "parameters": {"features": 512, "lr": 0.00011599497432385631, "momentum": 0.5394097683782016}, "parameter_index": 0}',
    index: 0
  },
  placementConstraint: { type: 'None', gpus: [] }
}
[2022-10-16 11:21:05] INFO (NNIManager) submitTrialJob: form: {
  sequenceId: 1,
  hyperParameters: {
    value: '{"parameter_id": 1, "parameter_source": "algorithm", "parameters": {"features": 128, "lr": 0.010676342217290587, "momentum": 0.5847482630995584}, "parameter_index": 0}',
    index: 0
  },
  placementConstraint: { type: 'None', gpus: [] }
}
[2022-10-16 11:21:24] INFO (NNIManager) Trial job CDKdy status changed from WAITING to RUNNING
[2022-10-16 11:21:37] INFO (NNIManager) Trial job Y3Rh9 status changed from WAITING to RUNNING
[2022-10-16 11:22:22] INFO (NNIManager) Trial job CDKdy status changed from RUNNING to SUCCEEDED
[2022-10-16 11:22:25] INFO (NNIManager) Trial job Y3Rh9 status changed from RUNNING to SUCCEEDED
[2022-10-16 11:22:25] INFO (NNIManager) NNIManager received command from dispatcher: TR, {"parameter_id": 2, "parameter_source": "algorithm", "parameters": {"features": 128, "lr": 0.01999772623507946, "momentum": 0.4162149739537774}, "parameter_index": 0}
[2022-10-16 11:22:25] INFO (NNIManager) NNIManager received command from dispatcher: TR, {"parameter_id": 3, "parameter_source": "algorithm", "parameters": {"features": 128, "lr": 0.0007597819239704861, "momentum": 0.6596009006777764}, "parameter_index": 0}
[2022-10-16 11:22:30] INFO (NNIManager) submitTrialJob: form: {
  sequenceId: 2,
  hyperParameters: {
    value: '{"parameter_id": 2, "parameter_source": "algorithm", "parameters": {"features": 128, "lr": 0.01999772623507946, "momentum": 0.4162149739537774}, "parameter_index": 0}',
    index: 0
  },
  placementConstraint: { type: 'None', gpus: [] }
}
[2022-10-16 11:22:30] INFO (NNIManager) submitTrialJob: form: {
  sequenceId: 3,
  hyperParameters: {
    value: '{"parameter_id": 3, "parameter_source": "algorithm", "parameters": {"features": 128, "lr": 0.0007597819239704861, "momentum": 0.6596009006777764}, "parameter_index": 0}',
    index: 0
  },
  placementConstraint: { type: 'None', gpus: [] }
}
[2022-10-16 11:22:42] INFO (NNIManager) Trial job WTp3h status changed from WAITING to RUNNING
[2022-10-16 11:22:52] INFO (NNIManager) Trial job G7Gqp status changed from WAITING to RUNNING
[2022-10-16 11:23:36] INFO (NNIManager) Trial job WTp3h status changed from RUNNING to SUCCEEDED
[2022-10-16 11:23:39] INFO (NNIManager) NNIManager received command from dispatcher: TR, {"parameter_id": 4, "parameter_source": "algorithm", "parameters": {"features": 256, "lr": 0.0007749163910847291, "momentum": 0.426688977511853}, "parameter_index": 0}
[2022-10-16 11:23:46] INFO (NNIManager) Trial job G7Gqp status changed from RUNNING to SUCCEEDED
[2022-10-16 11:23:46] INFO (NNIManager) submitTrialJob: form: {
  sequenceId: 4,
  hyperParameters: {
    value: '{"parameter_id": 4, "parameter_source": "algorithm", "parameters": {"features": 256, "lr": 0.0007749163910847291, "momentum": 0.426688977511853}, "parameter_index": 0}',
    index: 0
  },
  placementConstraint: { type: 'None', gpus: [] }
}
[2022-10-16 11:23:46] INFO (NNIManager) NNIManager received command from dispatcher: TR, {"parameter_id": 5, "parameter_source": "algorithm", "parameters": {"features": 512, "lr": 0.0005541282701768431, "momentum": 0.7236941232867435}, "parameter_index": 0}
[2022-10-16 11:23:46] INFO (NNIManager) submitTrialJob: form: {
  sequenceId: 5,
  hyperParameters: {
    value: '{"parameter_id": 5, "parameter_source": "algorithm", "parameters": {"features": 512, "lr": 0.0005541282701768431, "momentum": 0.7236941232867435}, "parameter_index": 0}',
    index: 0
  },
  placementConstraint: { type: 'None', gpus: [] }
}
[2022-10-16 11:23:58] INFO (NNIManager) Trial job AnbTs status changed from WAITING to RUNNING
[2022-10-16 11:24:09] INFO (NNIManager) Trial job G21j9 status changed from WAITING to RUNNING
[2022-10-16 11:24:54] INFO (NNIManager) Trial job AnbTs status changed from RUNNING to SUCCEEDED
[2022-10-16 11:24:56] INFO (NNIManager) NNIManager received command from dispatcher: TR, {"parameter_id": 6, "parameter_source": "algorithm", "parameters": {"features": 256, "lr": 0.005231829898637109, "momentum": 0.3158966535554839}, "parameter_index": 0}
[2022-10-16 11:25:03] INFO (NNIManager) Trial job G21j9 status changed from RUNNING to SUCCEEDED
[2022-10-16 11:25:03] INFO (NNIManager) submitTrialJob: form: {
  sequenceId: 6,
  hyperParameters: {
    value: '{"parameter_id": 6, "parameter_source": "algorithm", "parameters": {"features": 256, "lr": 0.005231829898637109, "momentum": 0.3158966535554839}, "parameter_index": 0}',
    index: 0
  },
  placementConstraint: { type: 'None', gpus: [] }
}
[2022-10-16 11:25:03] INFO (NNIManager) NNIManager received command from dispatcher: TR, {"parameter_id": 7, "parameter_source": "algorithm", "parameters": {"features": 512, "lr": 0.015826850876395914, "momentum": 0.032269555491401314}, "parameter_index": 0}
[2022-10-16 11:25:03] INFO (NNIManager) submitTrialJob: form: {
  sequenceId: 7,
  hyperParameters: {
    value: '{"parameter_id": 7, "parameter_source": "algorithm", "parameters": {"features": 512, "lr": 0.015826850876395914, "momentum": 0.032269555491401314}, "parameter_index": 0}',
    index: 0
  },
  placementConstraint: { type: 'None', gpus: [] }
}
[2022-10-16 11:25:16] INFO (NNIManager) Trial job Idgwf status changed from WAITING to RUNNING
 

 - dispatcher.log:

[2022-10-16 11:21:00] INFO (nni.tuner.tpe/MainThread) Using random seed 1981398779
[2022-10-16 11:21:00] INFO (nni.runtime.msg_dispatcher_base/MainThread) Dispatcher started

 - nnictl stdout and stderr:
 

**How to reproduce it?**:

`# Importing required libraries
from nni.experiment import Experiment

# Defining the hyperparameter search space
search_space = {
    'features': {'_type': 'choice', '_value': [128, 256, 512, 1024]},
    'lr': {'_type': 'loguniform', '_value': [0.0001, 0.1]},
    'momentum': {'_type': 'uniform', '_value': [0, 1]},
}

# Setting up the NNI experiment on local machine
experiment = Experiment('local')

# Conducting NNI evaluation in trail mode
experiment.config.trial_command = 'python model.py'
# experiment.config.trial_code_directory = '.'

# Configuring the search space
experiment.config.search_space = search_space

# Configuring the tuning algorithm
experiment.config.tuner.name = 'TPE'
experiment.config.tuner.class_args['optimize_mode'] = 'maximize'

# Setting up number of trials to run -> Sets of hyperparameters and trial concurrency
experiment.config.max_trial_number = 50  # Change to a higher number
experiment.config.trial_concurrency = 2

# Running the experiment on portal
experiment.run(8057)

# Stopping the experiment
experiment.stop()

# For still viewing the experiment
# experiment.view()
`

Run this with model.py file from https://github.com/microsoft/nni/tree/dab51f799f77aa72c18774faffaedf8d0ee2c977/examples/tutorials/hpo_quickstart_pytorch

**Comment by liuzhe-lz** [comment]
Please add `"debug": true` to experiment config and check `dispatcher.log`.
Also, monitor when does the dispatcher process (`python -m nni --exp_parms [long base64 string]`) ends (with Task Manager or `top`).

**Comment by imsaumil** [comment]
Please find the dispatcher log file : [dispatcher.log](https://github.com/microsoft/nni/files/9811332/dispatcher.log)

I am getting the same error with concurrency = 1.

**Comment by imsaumil** [comment]
Hi @liuzhe-lz, I was following the recent commit regarding letting the WebSocket reconnect. I tried upgrading nni and it gives the same error. 

**COMMAND PROMPT OUTPUT:**
[2022-10-19 11:14:39] Web portal URLs: http://10.76.56.148:8057 http://169.254.31.88:8057 http://169.254.194.255:8057 http://169.254.35.0:8057 http://127.0.0.1:8057
Error: Dispatcher stream error, tuner may have crashed.                                                                       
    at EventEmitter.<anonymous> (C:\Users\sashah8\Anaconda3\envs\nni_HPO\lib\site-packages\nni_node\core\nnimanager.js:647:32)
    at EventEmitter.emit (node:events:526:28)
    at WebSocketChannelImpl.handleError (C:\Users\sashah8\Anaconda3\envs\nni_HPO\lib\site-packages\nni_node\core\tuner_command_channel\websocket_channel.js:107:22)
    at WebSocketChannelImpl.heartbeat (C:\Users\sashah8\Anaconda3\envs\nni_HPO\lib\site-packages\nni_node\core\tuner_command_channel\websocket_channel.js:91:18)
    at listOnTimeout (node:internal/timers:559:17)
    at processTimers (node:internal/timers:502:7)
[2022-10-19 11:15:22] Stopping experiment, please wait...
[2022-10-19 11:15:22] Experiment stopped

**LOG FILES:**
1) [dispatcher.log](https://github.com/microsoft/nni/files/9821806/dispatcher.log)
2) [nnimanager.log](https://github.com/microsoft/nni/files/9821810/nnimanager.log)

**Comment by liuzhe-lz** [comment]
The last 2 log files are full log? The dispatcher log looks like being forcibly killed.

**Comment by Lijiaoa** [comment]
hi @imsaumil Do you still have this installation issue? Could you update it? Thanks!

**Comment by imsaumil** [comment]
Hi @Lijiaoa, I am still facing this issue but running it on a cluster as opposed to my local computer solved the problem. But I am still unable to figure out why does it not work on my local computer.

**Comment by Lijiaoa** [comment]
@imsaumil 2.10 will be released soon! and I suggest you to use v2.10, maybe this issue had been resolved in that version!
